src/core/queryCache.ts

```diff
@@ -20,7 +21,48 @@ interface QueryHashMap {
   [hash: string]: Query<any, any>
 }
 
-type QueryCacheListener = (query?: Query) => void
+interface NotifyEventQueryAdded {
+  type: 'queryAdded'
+  query: Query<any, any>
+}
+
+interface NotifyEventQueryRemoved {
+  type: 'queryRemoved'
+  query: Query<any, any>
+}
+
+interface NotifyEventObserverAdded {
+  type: 'observerAdded'
+  query: Query<any, any>
+  observer: QueryObserver<any, any, any, any>
+}
+
+interface NotifyEventObserverRemoved {
+  type: 'observerRemoved'
+  query: Query<any, any>
+  observer: QueryObserver<any, any, any, any>
+}
+
+interface NotifyEventUpdateResults {
+  type: 'updateResults'
+  query: Query<any, any>
+}
+
+interface NotifyEventQueryUpdated {
+  type: 'dispatch'
+  query: Query<any, any>
+  action: Action<any, any>
+}
+
+type QueryCacheNotifyEvent =
+  | NotifyEventQueryAdded
+  | NotifyEventQueryRemoved
+  | NotifyEventObserverAdded
+  | NotifyEventObserverRemoved
+  | NotifyEventUpdateResults
```

Comment:
Might be good to add `Observer` to the name like `NotifyEventQueryObserverUpdated` as this event is observer specific?